// @utxorpc/spec/src/utxorpc/v1alpha/sync/sync_connect.ts
// @generated by protoc-gen-connect-es v1.4.0 with parameter "target=ts"
// @generated from file utxorpc/v1alpha/sync/sync.proto (package utxorpc.v1alpha.sync, syntax proto3)
/* eslint-disable */
// @ts-nocheck

import { DumpHistoryRequest, DumpHistoryResponse, FetchBlockRequest, FetchBlockResponse, FollowTipRequest, FollowTipResponse, ReadTipRequest, ReadTipResponse } from "./sync_pb.js";
import { MethodKind } from "@bufbuild/protobuf";

/**
 * Service definition for syncing chain data.
 *
 * @generated from service utxorpc.v1alpha.sync.SyncService
 */
export const SyncService = {
  typeName: "utxorpc.v1alpha.sync.SyncService",
  methods: {
    /**
     * Fetch a block by its reference.
     *
     * @generated from rpc utxorpc.v1alpha.sync.SyncService.FetchBlock
     */
    fetchBlock: {
      name: "FetchBlock",
      I: FetchBlockRequest,
      O: FetchBlockResponse,
      kind: MethodKind.Unary,
    },
    /**
     * Dump the block history.
     *
     * @generated from rpc utxorpc.v1alpha.sync.SyncService.DumpHistory
     */
    dumpHistory: {
      name: "DumpHistory",
      I: DumpHistoryRequest,
      O: DumpHistoryResponse,
      kind: MethodKind.Unary,
    },
    /**
     * Follow the tip of the blockchain.
     *
     * @generated from rpc utxorpc.v1alpha.sync.SyncService.FollowTip
     */
    followTip: {
      name: "FollowTip",
      I: FollowTipRequest,
      O: FollowTipResponse,
      kind: MethodKind.ServerStreaming,
    },
    /**
     * Read the current tip of the blockchain.
     *
     * @generated from rpc utxorpc.v1alpha.sync.SyncService.ReadTip
     */
    readTip: {
      name: "ReadTip",
      I: ReadTipRequest,
      O: ReadTipResponse,
      kind: MethodKind.Unary,
    },
  }
} as const;
