// @utxorpc/spec/src/utxorpc/v1alpha/query/query_connect.ts
/// A consistent view of the state of the ledger

// @generated by protoc-gen-connect-es v1.4.0 with parameter "target=ts"
// @generated from file utxorpc/v1alpha/query/query.proto (package utxorpc.v1alpha.query, syntax proto3)
/* eslint-disable */
// @ts-nocheck

import { ReadDataRequest, ReadDataResponse, ReadParamsRequest, ReadParamsResponse, ReadUtxosRequest, ReadUtxosResponse, SearchUtxosRequest, SearchUtxosResponse } from "./query_pb.js";
import { MethodKind } from "@bufbuild/protobuf";

/**
 * Service definition for querying the state of the chain.
 *
 * @generated from service utxorpc.v1alpha.query.QueryService
 */
export const QueryService = {
  typeName: "utxorpc.v1alpha.query.QueryService",
  methods: {
    /**
     * Get overall chain state.
     *
     * @generated from rpc utxorpc.v1alpha.query.QueryService.ReadParams
     */
    readParams: {
      name: "ReadParams",
      I: ReadParamsRequest,
      O: ReadParamsResponse,
      kind: MethodKind.Unary,
    },
    /**
     * Read specific UTxOs by reference.
     *
     * @generated from rpc utxorpc.v1alpha.query.QueryService.ReadUtxos
     */
    readUtxos: {
      name: "ReadUtxos",
      I: ReadUtxosRequest,
      O: ReadUtxosResponse,
      kind: MethodKind.Unary,
    },
    /**
     * Search for UTxO based on a pattern.
     *
     * @generated from rpc utxorpc.v1alpha.query.QueryService.SearchUtxos
     */
    searchUtxos: {
      name: "SearchUtxos",
      I: SearchUtxosRequest,
      O: SearchUtxosResponse,
      kind: MethodKind.Unary,
    },
    /**
     * Read specific datum by hash
     *
     * @generated from rpc utxorpc.v1alpha.query.QueryService.ReadData
     */
    readData: {
      name: "ReadData",
      I: ReadDataRequest,
      O: ReadDataResponse,
      kind: MethodKind.Unary,
    },
  }
} as const;
